### PR TITLE
test: chainid, switch custom net, batch tx diff, snaps

### PR DESCRIPTION
**Pull request by seaona**
## **Description**

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27725?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: test/e2e/snaps/test-snap-txinsights-v2.spec.js [hunk 1]
```diff
@@ -37,22 +36,18 @@ describe('Test Snap TxInsights-v2', function () {
...
-
         await driver.clickElement({
           text: 'Confirm',
           tag: 'button',
         });
 
-        await driver.waitForSelector({ text: 'OK' });
-
-        await driver.clickElement({
+        await driver.clickElementAndWaitForWindowToClose({
```
Comment (seaona) [inline test/e2e/snaps/test-snap-txinsights-v2.spec.js:50]: this mitigates race conditions: we should not proceed to the next step until the dialog is closed

File: test/e2e/snaps/test-snap-txinsights-v2.spec.js [hunk 2]
```diff
@@ -62,8 +57,8 @@ describe('Test Snap TxInsights-v2', function () {
         await driver.clickElement('#getAccounts');
 
         // switch back to MetaMask window and deal with dialogs
-        await switchToNotificationWindow(driver);
-        await driver.clickElement({
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+        await driver.clickElementAndWaitForWindowToClose({
```
Comment (seaona) [inline test/e2e/snaps/test-snap-txinsights-v2.spec.js:61]: same, wait for the dialog to close before proceeding

File: test/e2e/snaps/test-snap-txinsights.spec.js [hunk 3]
```diff
@@ -37,22 +36,18 @@ describe('Test Snap TxInsights', function () {
...
 
         // switch to metamask extension and click connect
-        await switchToNotificationWindow(driver, 2);
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
         await driver.clickElement({
           text: 'Connect',
           tag: 'button',
         });
 
-        await driver.waitForSelector({ text: 'Confirm' });
```
Comment (seaona) [inline test/e2e/snaps/test-snap-txinsights.spec.js:46]: not needed

File: test/e2e/snaps/test-snap-txinsights.spec.js [hunk 4]
```diff
@@ -37,22 +36,18 @@ describe('Test Snap TxInsights', function () {
...
-
         await driver.clickElement({
           text: 'Confirm',
           tag: 'button',
         });
 
-        await driver.waitForSelector({ text: 'OK' });
-
-        await driver.clickElement({
+        await driver.clickElementAndWaitForWindowToClose({
```
Comment (seaona) [inline test/e2e/snaps/test-snap-txinsights.spec.js:50]: wait for the dialog to close before proceeding

File: test/e2e/snaps/test-snap-txinsights.spec.js [hunk 5]
```diff
@@ -62,8 +57,8 @@ describe('Test Snap TxInsights', function () {
         await driver.clickElement('#getAccounts');
 
         // switch back to MetaMask window and deal with dialogs
-        await switchToNotificationWindow(driver, 2);
-        await driver.clickElement({
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+        await driver.clickElementAndWaitForWindowToClose({
```
Comment (seaona) [inline test/e2e/snaps/test-snap-txinsights.spec.js:61]: wait for the dialog to close before proceeding

File: test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js [hunk 6]
```diff
@@ -49,14 +47,11 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks', fun
         await openDapp(driver, undefined, DAPP_URL);
 
         // Connect to dapp 1
-        await driver.findClickableElement({ text: 'Connect', tag: 'button' });
-        await driver.clickElement('#connectButton');
+        await driver.clickElement({ text: 'Connect', tag: 'button' });
 
-        await driver.delay(regularDelayMs);
```
Comment (seaona) [inline test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js:55]: not needed

File: test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js [hunk 7]
```diff
@@ -49,14 +47,11 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks', fun
...
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
 
-        await switchToNotificationWindow(driver);
-
-        await driver.clickElement({
+        await driver.clickElementAndWaitForWindowToClose({
```
Comment (seaona) [inline test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js:54]: wait for the dialog to close before proceeding

File: test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js [hunk 8]
```diff
@@ -82,14 +77,11 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks', fun
         await openDapp(driver, undefined, DAPP_ONE_URL);
 
         // Connect to dapp 2
-        await driver.findClickableElement({ text: 'Connect', tag: 'button' });
```
Comment (seaona) [inline test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js:85]: not needed

File: test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js [hunk 9]
```diff
@@ -82,14 +77,11 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks', fun
...
-        await driver.clickElement('#connectButton');
-
-        await driver.delay(regularDelayMs);
```
Comment (seaona) [inline test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js:88]: not needed

File: test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js [hunk 10]
```diff
@@ -108,24 +100,29 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks', fun
...
+        await driver.waitForSelector(
           By.xpath("//div[normalize-space(.)='1 of 2']"),
         );
 
         // Reject All Transactions
         await driver.clickElement('.page-container__footer-secondary a');
 
-        await driver.clickElement({ text: 'Reject all', tag: 'button' }); // TODO: Do we want to confirm here?
+        // TODO: Do we want to confirm here?
+        await driver.clickElementAndWaitForWindowToClose({
```
Comment (seaona) [inline test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js:113]: wait for the dialog to close before proceeding

File: test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js [hunk 11]
```diff
@@ -108,24 +100,29 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks', fun
...
-        await driver.waitUntilXWindowHandles(3);
+        // TODO: find a better way to handle different dialog ids
+        await driver.delay(2000);
 
         // Wait for new confirmations queued from second dapp to open
-        await switchToNotificationWindow(driver, 4);
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
 
-        await driver.findElement(
+        await driver.waitForSelector(
```
Comment (seaona) [inline test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js:125]: nit

File: test/e2e/tests/request-queuing/chainid-check.spec.js [hunk 12]
```diff
@@ -233,13 +233,11 @@ describe('Request Queueing chainId proxy sync', function () {
           assert.equal(chainIdBeforeConnectAfterManualSwitch, '0x1');
 
           // Connect to dapp
-          await driver.findClickableElement({ text: 'Connect', tag: 'button' });
-          await driver.clickElement('#connectButton');
+          await driver.clickElement({ text: 'Connect', tag: 'button' });
 
-          await driver.delay(regularDelayMs);
```
Comment (seaona) [inline test/e2e/tests/request-queuing/chainid-check.spec.js:239]: not needed

File: test/e2e/tests/request-queuing/chainid-check.spec.js [hunk 13]
```diff
@@ -280,6 +281,10 @@ describe('Request Queueing chainId proxy sync', function () {
           // should be on the new chainId that was requested
           assert.equal(chainIdAfterDappSwitch, '0x539'); // 1337
 
+          await driver.waitForSelector({
+            css: '[id="chainId"]',
+            text: '0x539',
+          });
```
Comment (seaona) [inline test/e2e/tests/request-queuing/chainid-check.spec.js:287]: ensure the test dapp has the new chain id too

File: test/e2e/tests/request-queuing/chainid-check.spec.js [hunk 14]
```diff
@@ -252,6 +250,10 @@ describe('Request Queueing chainId proxy sync', function () {
 
           // should still be on the same chainId as the wallet after connecting
           assert.equal(chainIdAfterConnect, '0x1');
+          await driver.waitForSelector({
+            css: '[id="chainId"]',
+            text: '0x1',
+          });
```
Comment (seaona) [inline test/e2e/tests/request-queuing/chainid-check.spec.js:256]: ensure the test dapp has the new chain id too

**Review by NidhiKJha — APPROVED** [review]
LGTM

**Pull request merged**